test/lit/passes/gufa.wast

```diff
@@ -0,0 +1,967 @@
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.get $x)
+  ;; CHECK-NEXT: )
+  (func $local-no (result i32)
+    (local $x i32)
+    (if
+      (call $import)
+      (local.set $x
+        (i32.const 1)
+      )
+    )
+    ;; $x has two possible values, 1 and the default 0, so we cannot optimize
+    ;; anything here.
+    (local.get $x)
+  )
+
+  ;; CHECK:      (func $local-yes (result i32)
+  ;; CHECK-NEXT:  (local $x i32)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result i32)
+  ;; CHECK-NEXT:    (if
+  ;; CHECK-NEXT:     (call $import)
+  ;; CHECK-NEXT:     (local.set $x
+  ;; CHECK-NEXT:      (i32.const 0)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (i32.const 0)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $local-yes (result i32)
+    (local $x i32)
+    (if
+      (call $import)
+      (local.set $x
+        ;; As above, but now we set 0 here. We can optimize the local.get to 0
+        ;; in this case.
+        (i32.const 0)
+      )
+    )
+    (local.get $x)
+  )
+
+  ;; CHECK:      (func $param-no (param $param i32) (result i32)
+  ;; CHECK-NEXT:  (if
+  ;; CHECK-NEXT:   (local.get $param)
+  ;; CHECK-NEXT:   (local.set $param
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (local.get $param)
+  ;; CHECK-NEXT: )
+  (func $param-no (export "param-no") (param $param i32) (result i32)
+    (if
+      (local.get $param)
+      (local.set $param
+        (i32.const 1)
+      )
+    )
+    ;; $x has two possible values, the incoming param value and 1, so we cannot
+    ;; optimize, since the function is exported - anything on the outside could
+    ;; call it with values we are not aware of during the optimization.
+    (local.get $param)
+  )
+
+  ;; CHECK:      (func $param-yes (param $param i32) (result i32)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result i32)
+  ;; CHECK-NEXT:    (if
+  ;; CHECK-NEXT:     (i32.const 1)
+  ;; CHECK-NEXT:     (local.set $param
+  ;; CHECK-NEXT:      (i32.const 1)
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:    (i32.const 1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.const 1)
+  ;; CHECK-NEXT: )
+  (func $param-yes (param $param i32) (result i32)
+    (if
+      (local.get $param)
+      (local.set $param
+        (i32.const 1)
+      )
+    )
+    ;; As above, but now the function is not exported. That means it has no
+    ;; callers, so the only possible contents for $param are the local.set here,
+    ;; as this code is unreachable. We will infer a constant of 1 for all values
+    ;; of $param here. (With an SSA analysis, we could see that the first
+    ;; local.get must be unreachable, and optimize even better; as things are,
+    ;; we see the local.set and it is the only thing that sends values to the
+    ;; local.)
```

Comment:
Without SSA we consider a local index either reachable or not as a whole, not individual sets and gets. This local index seems reachable due to the set here. But the set itself is not really reachable since it's behind an unreachable get - but we need an SSA-like analysis to see that.